compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -9528,8 +9528,11 @@ private void handleSafeNavigation(BLangAccessExpression accessExpr, BType type,
         pushToMatchStatementStack(matchStmt, successClause, pos);
     }
 
-    private boolean isMapJson(BType originalType) {
-        return originalType.tag == TypeTags.MAP && ((BMapType) originalType).getConstraint().tag == TypeTags.JSON;
+    private boolean isMapJson(BType originalType, boolean fromMap) {
+        if (originalType.tag == TypeTags.MAP) {
+            return isMapJson(((BMapType) originalType).getConstraint(), true);
+        }
+        return originalType.tag == TypeTags.JSON && fromMap;
```

Comment:
We can directly return
return (originalType.tag == TypeTags.MAP && isMapJson(((BMapType) originalType).getConstraint(), true)) || 
originalType.tag == TypeTags.JSON && fromMap